# To eBike or Not to eBike?

## Remove Nulls, Duplicates, and Other Errata

<a id=toc></a>
## Table of Contents

<ul>
    <li><a href=#01-import-packages>Import Packages</a>
    <li><a href=#02-load-dataset>Load Dataset and Check Properties</a>
    <li><a href=#03-clean-data>Remove Null and Duplicate Values</a>
    <li><a href=#04-norm-data>Normalize Categorical Data Types</a>
    <li><a href=#05-save-data>Save Cleaned Dataset</a>
</ul>

<a id=01-import-packages></a>
## Import Packages

First, import relevant packages:

In [2]:
# # Numerical and data
# import pandas as pd
# import numpy as np

# # Apache parquet files (to save space)
# import pyarrow as pa
# import pyarrow.parquet as pq

<a href=#toc>Back to the top</a>

<a id=02-load-dataset></a>
## Load Dataset and Check Properties

Next, load the .parquet file into a dataframe:

In [6]:
# CB_Data = pq.read_table('CitiBike_data/202106-202205-citibike-tripdata.parquet').to_pandas()

In [7]:
# CB_Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29032983 entries, 0 to 29032982
Data columns (total 10 columns):
 #   Column              Dtype  
---  ------              -----  
 0   rideable_type       object 
 1   started_at          object 
 2   ended_at            object 
 3   start_station_name  object 
 4   end_station_name    object 
 5   start_lat           float64
 6   start_lng           float64
 7   end_lat             float64
 8   end_lng             float64
 9   member_casual       object 
dtypes: float64(4), object(6)
memory usage: 2.2+ GB


In [8]:
# CB_Data.columns

Index(['rideable_type', 'started_at', 'ended_at', 'start_station_name',
       'end_station_name', 'start_lat', 'start_lng', 'end_lat', 'end_lng',
       'member_casual'],
      dtype='object')

In [9]:
# CB_Data.head()

rideable_type           started_at             ended_at  \
0    docked_bike  2021-06-01 23:12:34  2021-06-01 23:14:46   
1    docked_bike  2021-06-16 17:14:56  2021-06-16 17:29:15   
2    docked_bike  2021-06-07 19:41:55  2021-06-07 19:51:28   
3  electric_bike  2021-06-17 15:13:15  2021-06-17 15:33:25   
4  electric_bike  2021-06-18 08:27:03  2021-06-18 08:53:37   

          start_station_name            end_station_name  start_lat  \
0        Driggs Ave & N 9 St      Bayard St & Leonard St  40.718169   
1       Fulton St & Broadway       Mercer St & Spring St  40.711066   
2      Devoe St & Lorimer St   Manhattan Av & Leonard St  40.713352   
3        Driggs Ave & N 9 St  Greenwich Ave & Charles St  40.718169   
4  Graham Ave & Conselyea St        E 30 St & Park Ave S  40.715143   

   start_lng    end_lat    end_lng member_casual  
0 -73.955201  40.719156 -73.948854        member  
1 -74.009447  40.723627 -73.999496        casual  
2 -73.949103  40.720840 -73.948440        casual  
3 -73.955201  40.735238 -74.000271        member  
4 -73.944507  40.744449 -73.983035        member

Count total number of rides during that time frame (June 2021 through May 2022):

In [24]:
# no_rides = len(CB_Data)

<a href=#toc>Back to the top</a>

<a id=03-clean-data></a>
## Remove Null and Duplicate Values

Check for null values.

In [26]:
# # "Ghost bikes" coming in from unknown locations
# bike_ghos = CB_Data.loc[pd.isnull(CB_Data.start_station_name)]
# no_bike_ghos = len(bike_ghos)

# # Bikes that are lost, i.e. not docked at the end
# bike_lost = CB_Data.loc[pd.isnull(CB_Data.end_station_name)]
# no_bike_lost = len(bike_lost)

# # Bikes that are docked at the same station they are picked up, for joyride, rider changing mind, defective bike, etc.
# bike_joyr = CB_Data.loc[CB_Data.start_station_name == CB_Data.end_station_name]
# no_bike_joyr = len(bike_joyr)

# print(f'Total number of Citibike rides from June 2021 through May 2022: {no_rides}')
# print(f'Total number of "ghost bikes" in that time frame: {no_bike_ghos}')
# print(f'Total number of lost bikes in that time frame: {no_bike_lost}')
# print(f'Total number of bikes being docked at the same location in that time frame: {no_bike_joyr}')
# print(f'Total number of "dud rides" to be removed from data: {no_bike_ghos + no_bike_lost + no_bike_joyr}')

# print(f'Percentage of rides from June 2021 through May 2022 with bikes missing: \
#       {100*(no_bike_ghos + no_bike_lost)/no_rides}')
# print(f'Percentage of rides from June 2021 through May 2022 with bikes docked at the same location: \
#       {100*no_bike_joyr/no_rides}')
# print(f'Total number of "dud rides" to be removed from data: \
#       {100*(no_bike_ghos + no_bike_lost + no_bike_joyr)/no_rides}')

Total number of Citibike rides from June 2021 through May 2022: 29032983
Total number of "ghost bikes" in that time frame: 245
Total number of lost bikes in that time frame: 102300
Total number of bikes being docked at the same location in that time frame: 1549785
Total number of "dud rides" to be removed from data: 1652330
Percentage of rides from June 2021 through May 2022 with bikes missing:       0.35320173610820493
Percentage of rides from June 2021 through May 2022 with bikes docked at the same location:       5.338015043097707
Total number of "dud rides" to be removed from data:       5.691216779205912


To summarize:
<br>Total number of Citibike rides from June 2021 through May 2022: 29,032,983
<br>Total number of "ghost bikes" in that time frame: 245
<br>Total number of lost bikes in that time frame: 102,300
<br>Total number of bikes being docked at the same location in that time frame: 1,549,785
<br>Total number of "dud rides" to be removed from data: 1,652,330
<br>Percentage of rides from June 2021 through May 2022 with bikes missing:       0.35320173610820493
<br>Percentage of rides from June 2021 through May 2022 with bikes docked at the same location:       5.338015043097707
<br>Total number of "dud rides" to be removed from data:       5.691216779205912

Out of 29 million rides taken from June 2021 through May 2022, 1.55 million (5.7%) such "dud rides" shall be removed.

Now, eliminate these "dud rides" once and for all:

In [40]:
# dud_rides = bike_ghos.index.tolist() + bike_lost.index.tolist() + bike_joyr.index.tolist()

# CB_Data = CB_Data.drop(axis = 0, index = dud_rides)

In [41]:
# len(CB_Data)

27380897

Now, there are 27 million rides to work with, which is still more than enough data!

<a href=#toc>Back to the top</a>

<a id=04-norm-data></a>
## Normalize Categorical Data Types

Make sure that all the values of **rideable_type** are consistent. In other words, **docked_bike** = **classic_bike**.

In [42]:
# CB_Data.rideable_type.loc[CB_Data.rideable_type == 'classic_bike'] = 'Classic Bike'
# CB_Data.rideable_type.loc[CB_Data.rideable_type == 'docked_bike'] = 'Classic Bike'
# CB_Data.rideable_type.loc[CB_Data.rideable_type == 'electric_bike'] = 'Electric Bike'

/Users/jchatterjee/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [43]:
# CB_Data.head()

rideable_type           started_at             ended_at  \
0   Classic Bike  2021-06-01 23:12:34  2021-06-01 23:14:46   
1   Classic Bike  2021-06-16 17:14:56  2021-06-16 17:29:15   
2   Classic Bike  2021-06-07 19:41:55  2021-06-07 19:51:28   
3  Electric Bike  2021-06-17 15:13:15  2021-06-17 15:33:25   
4  Electric Bike  2021-06-18 08:27:03  2021-06-18 08:53:37   

          start_station_name            end_station_name  start_lat  \
0        Driggs Ave & N 9 St      Bayard St & Leonard St  40.718169   
1       Fulton St & Broadway       Mercer St & Spring St  40.711066   
2      Devoe St & Lorimer St   Manhattan Av & Leonard St  40.713352   
3        Driggs Ave & N 9 St  Greenwich Ave & Charles St  40.718169   
4  Graham Ave & Conselyea St        E 30 St & Park Ave S  40.715143   

   start_lng    end_lat    end_lng member_casual  
0 -73.955201  40.719156 -73.948854        member  
1 -74.009447  40.723627 -73.999496        casual  
2 -73.949103  40.720840 -73.948440        casual  
3 -73.955201  40.735238 -74.000271        member  
4 -73.944507  40.744449 -73.983035        member

Capitalize **member_casual** entries.

In [94]:
# CB_Data.member_casual.loc[CB_Data.member_casual == 'member'] = 'Member'
# CB_Data.member_casual.loc[CB_Data.member_casual == 'casual'] = 'Casual'

/Users/jchatterjee/opt/anaconda3/lib/python3.9/site-packages/pandas/core/indexing.py:1732: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_block(indexer, value, name)


In [95]:
# CB_Data.head()

rideable_type           started_at             ended_at  \
0   Classic Bike  2021-06-01 23:12:34  2021-06-01 23:14:46   
1   Classic Bike  2021-06-16 17:14:56  2021-06-16 17:29:15   
2   Classic Bike  2021-06-07 19:41:55  2021-06-07 19:51:28   
3  Electric Bike  2021-06-17 15:13:15  2021-06-17 15:33:25   
4  Electric Bike  2021-06-18 08:27:03  2021-06-18 08:53:37   

          start_station_name            end_station_name  start_lat  \
0        Driggs Ave & N 9 St      Bayard St & Leonard St  40.718169   
1       Fulton St & Broadway       Mercer St & Spring St  40.711066   
2      Devoe St & Lorimer St   Manhattan Av & Leonard St  40.713352   
3        Driggs Ave & N 9 St  Greenwich Ave & Charles St  40.718169   
4  Graham Ave & Conselyea St        E 30 St & Park Ave S  40.715143   

   start_lng    end_lat    end_lng member_casual  
0 -73.955201  40.719156 -73.948854        Member  
1 -74.009447  40.723627 -73.999496        Casual  
2 -73.949103  40.720840 -73.948440        Casual  
3 -73.955201  40.735238 -74.000271        Member  
4 -73.944507  40.744449 -73.983035        Member

<a href=#toc>Back to the top</a>

<a id=05-save-data></a>
## Save Cleaned Dataset

At this point, it is a good time to save the cleaned file in .parquet.

In [96]:
# CB_Data_arrow = pa.Table.from_pandas(CB_Data)
# pq.write_table(CB_Data_arrow, 'CitiBike_data/202106-202205-citibike-tripdata.parquet')

<a href=#toc>Back to the top</a>